### PR TITLE
Abstract kvdb with cache

### PR DESCRIPTION
**Pull request by ordian**
The caching logic for kvdb is orthogonal to rocksdb and can be reused by other kvdb implementations.
Notable changes:
* iterator now acquires read lock on the db, which prevents error in case of db being closed in a separate thread (for parity-ethereum it can happen after warp sync is complete). The downside is if `close` is called on the same thread as `iter`, it will deadlock (`let _ = rwlock.read(); let _ = rwlock.write()`).
* db transactions now allocate (`Box`), not sure if it's possible to avoid that (see TODOs for reviewers).

### CODE REVIEW

File: kvdb/src/lib.rs [hunk 1]
```diff
@@ -167,9 +181,432 @@ pub trait KeyValueDB: Sync + Send {
...
+}
+
+/// An abstraction over a concrete database read-only transaction implementation.
+pub trait ReadTransaction {
+	/// Get value by key.
+	fn get(self: Box<Self>, col: usize, key: &[u8]) -> io::Result<Option<DBValue>>;
+}
+
+pub trait TransactionHandler {
+	// TODO: how to avoid boxing?
```
Comment (dvdplm) [inline kvdb/src/lib.rs]: This would require [existential types](https://rust-lang.github.io/rfcs/2071-impl-trait-existential-types.html) right?

Comment (ordian) [reply kvdb/src/lib.rs]: Probably, also by fixing https://github.com/rust-lang/rust/issues/51386

Comment (dvdplm) [reply kvdb/src/lib.rs]: …so I guess that for the time being boxing is as good as it gets?

Comment (ordian) [reply kvdb/src/lib.rs]: I was hoping someone could with a better design, which would remove this requirement. Also I don't know impactful this boxing is.

**Comment by dvdplm** [comment]
@ordian is this still relevant, i.e. should still be reviewed?

**Comment by ordian** [comment]
@dvdplm this could be useful for lmdb and [sled](https://github.com/paritytech/parity-common/compare/ao-kvdb-sled) kvdb impls, but I can update it next year.

**Comment by ordian** [comment]
Let's close it for now.

**Pull request closed without merging**